Packet implementation for portability over binary transport protocols

matheo-lucak opened the pull request:
There is around 10 complete unit tests, ready to be reviewed

matheo-lucak committed:
[ADD] Base of packet implementation. All basics functions are made. Need to implement read and write functions for each types

[ADD] Simple criterion test. Needing more

matheo-lucak committed:
[ADD] packet_can_read + tests

matheo-lucak committed:
[ADD] Read and Write Packet functions for u-int8-16-32 + Tests

matheo-lucak committed:
[UPDATE] packet_resize now reallocs with size + 1, to let assure null terminated byte for strings

matheo-lucak committed:
[ADD] Packet: read and write functions. + Tests on string and on multiple type combinations

matheo-lucak committed:
[ADD] packer_read_until: Fully functionnal. Well tested.

[DEL] -g CFLAGS on compilation

matheo-lucak committed:
[FIX] packet_read_string_until is fixed: Place a \0 at the end of a string, and correctly moving the read_pos index. Adding some complete tests

theopomies commented on src/packet/read.c:
```diff
@@ -0,0 +1,29 @@
+        return PACKET_ERROR;
+    }
+    if (packet_can_read(packet, size) == false) {
+        set_error("packet_read: Not enought data to read.");
```
No T here -> enough

theopomies commented on include/packet.h:
```diff
@@ -12,48 +12,61 @@
     size_t size;
     size_t read_pos;
-    size_t send_pos;
     bool is_valid;
```
Not necessary now that packet_status_t conveys the info

theopomies commented on src/packet/read_string.c:
```diff
@@ -0,0 +1,47 @@
+        return status;
+    packet->data[packet->size] = '\0';
+    size_to_read = strlen(packet->data + packet->read_pos) + 1;
+    *data = malloc(size_to_read);
```
realloc instead of malloc, if *data is NULL it is equivalent else it is better

theopomies commented on src/packet/read_string_until.c:
```diff
@@ -0,0 +1,70 @@
+    packet_read_string_cut_on_delimiter(packet, delim, &size_to_read);
+    if (size_to_read == 0)
+        return PACKET_END_OF_PACKET;
+    *data = calloc(1, size_to_read + 1);
```
Same, reallocarray

theopomies reviewed (changes requested):
Please also add newline at EOF, after that it LGTM

matheo-lucak committed:
[DEL] Usless member variable of Packet: is_valid

matheo-lucak committed:
[ADD] tcp_send_packet sends the actual data stored in once

[UPDATE] packet_read_string: uses realloc to free given data if previously allocated.
